### ISSUE TITLE
Unable to start Sarracenia on Windows

### ISSUE THREAD

**Issue by yraheja77** [MetPX/sarracenia#898]
I am unable to start Sarracenia on Windows 11 or Windows Server. I have tried installing Sarracenia both by using the Windows installer files and by using Pip but somehow it doesn't download anything. The screenshot of log and configuration file is attached for reference. What am I doing wrong?
the .conf file has been converted to .txt to share on Github

[subscribe_dd_all_01.log](https://github.com/MetPX/sarracenia/files/13969235/subscribe_dd_all_01.log)
[subscribe_dd_all_02.log](https://github.com/MetPX/sarracenia/files/13969236/subscribe_dd_all_02.log)
[subscribe_dd_all_03.log](https://github.com/MetPX/sarracenia/files/13969237/subscribe_dd_all_03.log)
[subscribe_dd_all_04.log](https://github.com/MetPX/sarracenia/files/13969238/subscribe_dd_all_04.log)
[subscribe_dd_all_05.log](https://github.com/MetPX/sarracenia/files/13969239/subscribe_dd_all_05.log)
[dd_all - Copy.txt](https://github.com/MetPX/sarracenia/files/13969260/dd_all.-.Copy.txt)
![screenshot](https://github.com/MetPX/sarracenia/assets/42151924/4976f76b-71a6-4e5d-a28c-157fcd798d01)

**Comment by petersilva** [comment]
try adding:

```

tlsRigour lax

```
to the configuration file.  It is some sort of problem with SSL cert validation.  I'm a bit puzzled about the source, but disabling cert verification should help us move to the next step.

**Comment by yraheja77** [comment]
Hi,
added that but it doesnt work. Its some other issue.

Have I configured the directory correctly. I have mentioned
"directory mnt/C/MSCDatamartClone/dd_all2".

Regards,
Yogesh

**Comment by petersilva** [comment]
It is interesting that the directory option uses forward slashes and starts with mnt/C ... Are you running in WSL ?  For us, that's not windows, but an ordinary linux install, nothing about a windows install would apply.  We need to confirm the platform... mnt/C would be something we would see in WSL or some other container type arrangement.  if you are using WSL (Windows subsystem for linux) can you tell me which distribution of linux you are using (our favorite is ubuntu 22 fwiw, if you have something else we can adjust.)

**Comment by petersilva** [comment]
otoh, if you really are in windows, then the directory line should look something like:
```
directory c:\MSCDatamart\Clone\dd_all2

```

**Comment by petersilva** [comment]
fwiw, if you are able to run in WSL, it's definitely our preference.  for us, raw windows is the last choice.

**Comment by petersilva** [comment]
if on WSL, the directory would need to start with a / otherwise it's relative to wherever you are running...

directory /mnt/C/MSCDatamartClone/dd_all2

**Comment by yraheja77** [comment]
It works on WSL but I was hoping I could install it using Windows as Ubuntu is not supported by my organization. But let me try and get permissions for the same.

Can you also please share a sample config and python script that is run once a weather alert is issued by ECCC.

**Comment by petersilva** [comment]
subscribe/dd_all is a configuration to download absolutely everything, including radar, and satellite, and numericla model outputs... it's kind of a lot to work with and clean out before you can work with what you want. 

I suggest it's easier to play with a feed like dd_amis... which is just bulletins, but they are continuous, so you know it is working right away.   Warnings are infrequent, making them hard to debug with.

to get the dd_amis configuration.  I do::

    sr3 add subscribe\dd_amis

You need to edit the configuration, just like before to change the destination
directory.   To add a python plugin to it, you need to know where to put the code so sr3 will find it.  Use *sr3 features* for that.

```

PS C:\Users\peter> sr3 features
2024-01-21 19:21:17,159 2816 [INFO] sarracenia.flow loadCallbacks flowCallback plugins to load: ['sarracenia.flowcb.gather.message.Message', 'sarracenia.flowcb.retry.Retry', 'sarracenia.flowcb.housekeeping.resources.Resources', 'run', 'log']
2024-01-21 19:21:17,159 2816 [INFO] sarracenia.flowcb.log __init__ subscribe initialized with: logEvents: {'after_post', 'on_housekeeping', 'after_accept', 'after_work'},  logMessageDump: False

Status:    feature:   python imports:      Description:
Installed  amqp       amqp                 can connect to rabbitmq brokers
Installed  appdirs    appdirs              place configuration and state files appropriately for platform (windows/mac/linux)
Installed  filetypes  magic                able to set content headers
Installed  ftppoll    dateparser,pytz      able to poll with ftp
Installed  humanize   humanize,humanfriendly humans numbers that are easier to read.
Installed  mqtt       paho.mqtt.client     can connect to mqtt brokers
Installed  process    psutil               can monitor, start, stop processes:  Sr3 CLI should basically work
Absent     redis      redis,redis_lock     cannot use redis implementations of retry and nodupe
Installed  retry      jsonpickle           can write messages to local queues to retry failed publishes/sends/downloads
Installed  sftp       paramiko             can use sftp or ssh based services
Absent     vip        netifaces            will not be able to use the vip option for high availability clustering
Installed  watch      watchdog             watch directories
Absent     xattr      xattr                unable to store additional file metadata in extended attributes
Installed  ctypes     ctypes               will store file metadata in extended attributes

 state dir: C:\Users\peter\AppData\Local\MetPX\sr3\Cache
 config dir: C:\Users\peter\AppData\Local\MetPX\sr3

PS C:\Users\peter>
```

the last line is about a "config dir" which is where the configurations are stored.
You need to make a plugins directory under there.  From this example:

```
PS C:\Users\peter> cd  C:\Users\peter\AppData\Local\MetPX\sr3
PS C:\Users\peter\AppData\Local\MetPX\sr3> mkdir plugins


    Directory: C:\Users\peter\AppData\Local\MetPX\sr3


Mode                 LastWriteTime         Length Name
----                 -------------         ------ ----
d-----        2024-01-21   4:33 PM                plugins

PS C:\Users\peter\AppData\Local\MetPX\sr3> cd plugins

```

Then create a flow callback plugin from there.   for example:

```


PS C:\Users\peter\AppData\Local\MetPX\sr3\plugins> wget https://raw.githubusercontent.com/MetPX/sarracenia/development/sarracenia/flowcb/run.py -Outfile run.py

```

This is to download from github one of the built-in plugins... run.py allows the user to run arbitrary scripts.  Since it's built-in, you probably want to rename it to avoid confusion:

* mv run.py run2.py
* edit the file, and change the class name from "Class Run(FlowCB)" to "Class Run2(FlowCB)

to change what happens to files once they are downloaded, you can modify the after_work() routine.  The filename downloaded is already built in that routine.

to get it called from your configuration, edit your configuration file, adding the line:

```
batch 5

callback run2

```

The default batch size is something like 100 ? And it's hard to follow things during development, so recommend reducing it to a manageable number like 5.

The *callback* line makes the subscribe process look for and load your custom python.

The python in the windows binaries is python 3.10.11... 

To test this, I added another line to the configuration (sr3 edit subsrcibe\dd_amis ) :

```
run_work_item powershell.exe c:\Users\peter\hello_world.ps1

```

I created a powershell script *hell_world.ps1* to test the run routine, like so:

hello_world.ps1:
```
param (
    [string]$file = "input_file"
)
#$wh =  $file | Write-Host
Write-output 'Hello, World! input file is $($file)' 

```

to demonstrate normal use of the run.py plugin.  here is a sample run:

```

.2024-01-21 19:48:37,459 [INFO] 8108 sarracenia.flow loadCallbacks flowCallback plugins to load: ['sarracenia.flowcb.gather.message.Message', 'sarracenia.flowcb.retry.Retry', 'sarracenia.flowcb.housekeeping.resources.Resources', 'run', 'log']
2024-01-21 19:48:37,475 [INFO] 8108 sarracenia.flowcb.log __init__ subscribe initialized with: logEvents: {'after_accept', 'on_housekeeping', 'after_post', 'after_work'},  logMessageDump: False
2024-01-21 19:48:37,475 [INFO] 8108 sarracenia.flow run callbacks loaded: ['sarracenia.flowcb.gather.message.Message', 'sarracenia.flowcb.retry.Retry', 'sarracenia.flowcb.housekeeping.resources.Resources', 'run', 'log']
2024-01-21 19:48:37,475 [INFO] 8108 sarracenia.flow run pid: 8108 subscribe/dd_amis instance: 0
2024-01-21 19:48:39,081 [INFO] 8108 sarracenia.moth.amqp _queueDeclare queue declared q_anonymous_subscribe.dd_amis.60456605.48223788 (as: amqps://anonymous@dd.weather.gc.ca/), (messages waiting: 622)

```

this startup log messages show:
* the callback was loaded,
* the process connected to the broker,
* declared a queue, and the corresponding bindings.

After a few seconds, some bulletins should start downloading:

```

2024-01-21 19:56:31,211 [INFO] 16300 sarracenia.flowcb.log after_accept accepted: (lag: 819.25 )  exchange: xpublic subtopic: bulletins.alphanumeric.20240122.UQ.CYQD.00 a file with baseUrl: https://dd4.weather.gc.ca relPath: /bulletins/alphanumeric/20240122/UQ/CYQD/00/UQCN01_CYQD_220000__71867_49319
2024-01-21 19:56:31,211 [INFO] 16300 sarracenia.flowcb.log after_accept accepted: (lag: 813.74 )  exchange: xpublic subtopic: bulletins.alphanumeric.20240122.SX.KWAL.00 a file with baseUrl: https://dd5.weather.gc.ca relPath: /bulletins/alphanumeric/20240122/SX/KWAL/00/SXAK50_KWAL_220042___43883
2024-01-21 19:56:31,211 [INFO] 16300 sarracenia.flowcb.log after_accept accepted: (lag: 812.24 )  exchange: xpublic subtopic: bulletins.alphanumeric.20240122.SR.KWAL.00 a file with baseUrl: https://dd5.weather.gc.ca relPath: /bulletins/alphanumeric/20240122/SR/KWAL/00/SRCN40_KWAL_220042___60904
2024-01-21 19:56:31,211 [INFO] 16300 sarracenia.flowcb.log after_accept accepted: (lag: 812.24 )  exchange: xpublic subtopic: bulletins.alphanumeric.20240122.SR.KWAL.00 a file with baseUrl: https://dd5.weather.gc.ca relPath: /bulletins/alphanumeric/20240122/SR/KWAL/00/SRCN40_KWAL_220042___63610
2024-01-21 19:56:31,211 [INFO] 16300 sarracenia.flowcb.log after_accept accepted: (lag: 813.58 )  exchange: xpublic subtopic: bulletins.alphanumeric.20240122.SR.KWAL.00 a file with baseUrl: https://dd5.weather.gc.ca relPath: /bulletins/alphanumeric/20240122/SR/KWAL/00/SRCN40_KWAL_220042___58727

```

So the above indicates some messages were received about files being available, but the files are not downloaded yet.  If you want to change the names of the files so they are downloaded to different names, for example, then you adjust the msg['new_file'] field in the message.

You can see the it collects five messages in a group... that's the batch size. It hasn't downloaded anything yet.   You can also modify the after_accept routine
to move messages from worklist.incoming (the messages for files to download) to worklist.rejected, to further refine the product selection prior to download.

anyways, now it has a batch of messages for files to download, it does the "work" phase of processing downloading each file in turn.  To process the files after downloading, modify the *after_work* plugin..

For the given unmodified example (just with a parameter for the shell script)
below is output you can see in the log file:

```

Hello, World! input file is  C:\Users\Peter\TEMP\dd_amis\SRMT60_KWAL_220109___25130
2024-01-21 20:23:29,885 [INFO] 9084 run2 after_work ran ['powershell.exe', 'c:\\Users\\peter\\hello_world.ps1', 'C:\\Users\\Peter\\TEMP\\dd_amis\\SRMT60_KWAL_220109___25130'] successfully
Hello, World! input file is  C:\Users\Peter\TEMP\dd_amis\SNVD01_KWBC_212200_RRX__33839
2024-01-21 20:23:31,243 [INFO] 9084 run2 after_work ran ['powershell.exe', 'c:\\Users\\peter\\hello_world.ps1', 'C:\\Users\\Peter\\TEMP\\dd_amis\\SNVD01_KWBC_212200_RRX__33839'] successfully
Hello, World! input file is  C:\Users\Peter\TEMP\dd_amis\SNVD80_KWBC_220100___55218
2024-01-21 20:23:32,476 [INFO] 9084 run2 after_work ran ['powershell.exe', 'c:\\Users\\peter\\hello_world.ps1', 'C:\\Users\\Peter\\TEMP\\dd_amis\\SNVD80_KWBC_220100___55218'] successfully
Hello, World! input file is  C:\Users\Peter\TEMP\dd_amis\ULUS50_KWBC_220000_RRF__11262
2024-01-21 20:23:33,690 [INFO] 9084 run2 after_work ran ['powershell.exe', 'c:\\Users\\peter\\hello_world.ps1', 'C:\\Users\\Peter\\TEMP\\dd_amis\\ULUS50_KWBC_220000_RRF__11262'] successfully
Hello, World! input file is  C:\Users\Peter\TEMP\dd_amis\SRCN40_KWAL_220109___25307
2024-01-21 20:23:34,902 [INFO] 9084 run2 after_work ran ['powershell.exe', 'c:\\Users\\peter\\hello_world.ps1', 'C:\\Users\\Peter\\TEMP\\dd_amis\\SRCN40_KWAL_220109___25307'] successfully

```

**Comment by yraheja77** [comment]
Thanks, I'll check it out.

**Comment by yraheja77** [comment]
Thank you Peter for the detailed explanation.
What do you recommend - WSL or containers in Windows for the production running of the tool?
Regards,
Yogesh

**Comment by petersilva** [comment]
All of our operations are on Linux, we don't have any experience with operational deployments on windows. The pure Windows deployment has an issue at the end of each day in that it cannot rollover it's logs, so, as per #785 you need a scheduled task to stop it, and then run sr3_rotateLogsManually once a day.

WSL (with ubuntu 22.04) is the closest thing to what we run in ops, so likely fewest issues.

**Comment by petersilva** [comment]
inactive... seemingly resolved...
